test/Replace.lean

```diff
@@ -0,0 +1,47 @@
+/-
+Copyright (c) 2022 Arthur Paulino. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arthur Paulino
+-/
+import Mathlib.Tactic.Replace
+
+-- tests with a explicitly named hipothesis
+
+example (h : Int) : Nat := by
+  replace h : Nat := 0
+  exact h
+
+example (h : Nat) : Nat := by
+  have h : Int := 0
+  assumption -- original `h` is not absent but...
+
+example (h : Nat) : Nat := by
+  replace h : Int := 0
+  fail_if_success assumption -- original `h` is absent now
+  replace h : Nat := 0
+  exact h
+
+-- tests with `this`
+
+example : Nat := by
+  have : Int := 0
+  replace : Nat := 0
+  assumption
+
+example : Nat := by
+  have : Nat := 0
+  have : Int := 0
+  assumption -- original `this` is not absent but...
+
+example : Nat := by
+  have : Nat := 0
+  replace : Int := 0
+  fail_if_success assumption -- original `this` is absent now
+  replace : Nat := 0
+  assumption
+
+-- trying to replace the type of a variable when the goal depends on it
+
+example {a : Nat} : a = a := by
+  fail_if_success replace a : Int := 0 -- tactic 'clear' failed, target depends on 'a'
```

Comment:
Did you check the lean 3 behavior? I think it just doesn't clear anything in this case.

Reply:
The Lean 3 behavior is indeed to fail:
```
import tactic.interactive

example {a : nat} : a = a :=
begin
  replace a : int := 1,
  -- 5:3: clear tactic failed, target type depends on 'a'
  -- state:
  -- a : ℕ,
  -- a : ℤ
  -- ⊢ a = a
end
```